fzf-tab.zsh

```diff
@@ -308,24 +308,29 @@ _fzf_tab_complete() {
 
 zle -C _fzf_tab_complete complete-word _fzf_tab_complete
 
+# note: this function is called with user options
 _fzf_tab_try_custom_completion() {
     # do not steal fzf's completions
-    [[ $LBUFFER =~ ${(q)FZF_COMPLETION_TRIGGER-'**'}$ ]] && return 1
-    local tokens=(${(z)LBUFFER})
-    [[ ${LBUFFER[-1]} = ' ' ]] && tokens+=("")
-    local cmd=${tokens[1]}
-    if (( $+functions[${FZF_TAB_CUSTOM_COMPLETIONS_PREFIX}${cmd}] )); then
-        local prefix=${tokens[-1]}
-        local lbuf
-        [ -z "${tokens[-1]}" ] && lbuf=$LBUFFER || lbuf=${LBUFFER:0:-${#tokens[-1]}}
-        prefix="$prefix" eval _fzf_complete_${cmd} ${(q)lbuf}
-        return 0
-    fi
-    return 1
+    [[ $LBUFFER != *"${FZF_COMPLETION_TRIGGER-**}" ]] || return
+    local func prefix lbuf
+    () {
+        emulate -L zsh
```

Comment:
Can I ask what is the benefit of `emulate -L zsh` in cases like this?

Reply:
The behavior of zsh can be changed with options. For example, `setopt ksh_arrays` has the following effects:

1. Arrays and scalars are now 0-based.
2. `$#foo` and `${#foo}` expand to the length of the first element in the array instead of the length of the array itself.
3. A few other things.

It's extremely difficult to write code that behaves correctly with all possible options, which on the surface would make it impossible to define functions that everyone could use regardless of their preferred options. To solve this problem zsh has builtin `emulate` that allows you to set all options to their default values. You can use `emulate sh` to make zsh behave like `sh` (some users actually put that in their `zshrc`), `emulate zsh` to make it behave "in native mode", and a few other modes. However, if you put `emulate zsh` in a function, you'll override options the user chose for themselves. Perhaps they put `setopt extended_glob` in `zshrc` but after calling the function this option is gone. To solve *this* problem there is `local_options` option, which causes options to be restored when function returns. `emulate -L zsh` is like `emulate zsh` plus `local_options` (and a few other `local_*` things).

Writing zsh functions that work for everyone is *really difficult*. In comparison, writing zsh scripts is relatively straightforward.